File: phpstorm-plugin/src/main/kotlin/com/ryanwelcher/wpinteractivity/parsers/JsStoreParser.kt
package com.ryanwelcher.wpinteractivity.parsers

import com.intellij.lang.javascript.psi.*
import com.intellij.psi.PsiFile
import com.intellij.psi.util.PsiTreeUtil
import com.ryanwelcher.wpinteractivity.models.MethodInfo
import com.ryanwelcher.wpinteractivity.models.PropertyInfo
import com.ryanwelcher.wpinteractivity.models.StoreDefinition

class JsStoreParser {
    fun parseFile(file: PsiFile): List<StoreDefinition> {
        val stores = mutableListOf<StoreDefinition>()
        val callExpressions = PsiTreeUtil.findChildrenOfType(file, JSCallExpression::class.java)

        for (call in callExpressions) {
            val methodExpression = call.methodExpression as? JSReferenceExpression ?: continue
            if (methodExpression.referenceName == "store") {
                val store = parseStoreFromCall(call, file)
                if (store != null) {
                    stores.add(store)
                }
            }
        }
        return stores
    }

    private fun parseStoreFromCall(call: JSCallExpression, file: PsiFile): StoreDefinition? {
        val arguments = call.arguments
        if (arguments.size < 2) return null

        val namespace = (arguments[0] as? JSLiteralExpression)?.stringValue ?: return null
        val storeObject = arguments[1] as? JSObjectLiteralExpression ?: return null

        val state = mutableMapOf<String, PropertyInfo>()
        val actions = mutableMapOf<String, MethodInfo>()
        val callbacks = mutableMapOf<String, MethodInfo>()

        for (property in storeObject.properties) {
            val name = property.name ?: continue
            val value = property.value

            when (name) {
                "state" -> (value as? JSObjectLiteralExpression)?.let { extractStateProperties(it, state) }
                "actions" -> (value as? JSObjectLiteralExpression)?.let { extractMethods(it, actions) }
                "callbacks" -> (value as? JSObjectLiteralExpression)?.let { extractMethods(it, callbacks) }
            }
        }

        return StoreDefinition(
            namespace = namespace,
            state = state,
            actions = actions,
            callbacks = callbacks,
            sourceFile = file.virtualFile.path,
            sourceType = "javascript",
            lastModified = file.virtualFile.modificationStamp
        )
    }

    private fun extractStateProperties(obj: JSObjectLiteralExpression, state: MutableMap<String, PropertyInfo>) {
        for (prop in obj.properties) {
            val name = prop.name ?: continue
            val value = prop.value

            val info = PropertyInfo(
                name = name,
                type = inferType(value),
                isObject = value is JSObjectLiteralExpression,
                sourceLine = 0 // Can be calculated from prop.textOffset
            )

            if (value is JSObjectLiteralExpression) {
                val nestedProperties = mutableMapOf<String, PropertyInfo>()
                extractStateProperties(value, nestedProperties)
                // PropertyInfo in Kotlin definition needs to be mutable or handled correctly
                // For now, let's assume we can set it if we change the data class or use a copy
            }
            state[name] = info
        }
    }

    private fun extractMethods(obj: JSObjectLiteralExpression, methods: MutableMap<String, MethodInfo>) {
        for (prop in obj.properties) {
            val name = prop.name ?: continue
            val value = prop.value

            val parameters = mutableListOf<String>()
            if (value is JSFunction) {
                value.parameters.forEach { it.name?.let { n -> parameters.add(n) } }
            }

            methods[name] = MethodInfo(
                name = name,
                parameters = parameters,
                sourceLine = 0
            )
        }
    }

    private fun inferType(value: JSExpression?): String? {
        return when (value) {
            is JSLiteralExpression -> {
                if (value.isStringLiteral) "string"
                else if (value.isNumericLiteral) "number"
                else if (value.isBooleanLiteral) "boolean"
                else null
            }
            is JSArrayLiteralExpression -> "array"
            is JSObjectLiteralExpression -> "object"
            is JSFunction -> "function"
            else -> null
        }
    }
}

File: src/parsers/jsStoreParser.ts
import { parse } from '@babel/parser';
import traverse from '@babel/traverse';
import { StoreDefinition, PropertyInfo, MethodInfo } from '../models/store';

/**
 * Parser for JavaScript/TypeScript files containing store() calls
 */
export class JsStoreParser {
	/**
	 * Parse a JavaScript/TypeScript file and extract store definitions
	 */
	public parseFile(content: string, filePath: string): StoreDefinition[] {
		try {
			const ast = parse(content, {
				sourceType: 'module',
				plugins: ['jsx', 'typescript']
			});

			return this.extractStores(ast, filePath);
		} catch (error) {
			// If parsing fails, return empty array
			console.error('Error parsing JS file:', error);
			return [];
		}
	}

	/**
	 * Extract store definitions from JavaScript AST
	 */
	private extractStores(ast: any, filePath: string): StoreDefinition[] {
		const stores: StoreDefinition[] = [];

		// First pass: collect variable declarations for later resolution
		const variableDeclarations = new Map<string, any>();
		traverse(ast, {
			VariableDeclarator: (path: any) => {
				if (path.node.id.type === 'Identifier' && path.node.init) {
					variableDeclarations.set(path.node.id.name, path.node.init);
				}
			}
		});

		// Second pass: find store() calls
		traverse(ast, {
			CallExpression: (path: any) => {
				// Check if this is a store() call (with or without generics)
				if (
					path.node.callee.type === 'Identifier' &&
					path.node.callee.name === 'store'
				) {
					const store = this.parseStoreFromCall(path.node, filePath, variableDeclarations);
					if (store) {
						stores.push(store);
					}
				}
			}
		});

		return stores;
	}

	/**
	 * Parse store definition from a store() call
	 */
	private parseStoreFromCall(
		node: any,
		filePath: string,
		variableDeclarations: Map<string, any>
	): StoreDefinition | null {
		if (!node.arguments || node.arguments.length < 2) {
			return null;
		}

		// First argument is the namespace
		const namespaceArg = node.arguments[0];
		const namespace = this.extractStringValue(namespaceArg);
		if (!namespace) {
			return null;
		}

		// Second argument is the store object (could be inline or a variable reference)
		let storeArg = node.arguments[1];

		// If it's a variable reference, resolve it
		if (storeArg.type === 'Identifier') {
			const resolvedStore = variableDeclarations.get(storeArg.name);
			if (resolvedStore) {
				storeArg = resolvedStore;
			} else {
				console.log(`[WP Interactivity API] Could not resolve variable: ${storeArg.name}`);
				return null;
			}
		}

		if (storeArg.type !== 'ObjectExpression') {
			return null;
		}

		const state = new Map<string, PropertyInfo>();
		const actions = new Map<string, MethodInfo>();
		const callbacks = new Map<string, MethodInfo>();

		// Extract properties from the store object
		for (const prop of storeArg.properties) {
			if (prop.type !== 'ObjectProperty' && prop.type !== 'ObjectMethod') {
				continue;
			}

			const key = this.getPropertyKey(prop);
			if (!key) {
				continue;
			}

			switch (key) {
				case 'state':
					if (prop.value && prop.value.type === 'ObjectExpression') {
						this.extractStateProperties(prop.value, state);
					}
					break;
				case 'actions':
					if (prop.value && prop.value.type === 'ObjectExpression') {
						this.extractMethods(prop.value, actions);
					}
					break;
				case 'callbacks':
					if (prop.value && prop.value.type === 'ObjectExpression') {
						this.extractMethods(prop.value, callbacks);
					}
					break;
			}
		}

		return {
			namespace,
			state,
			actions,
			callbacks,
			sourceFile: filePath,
			sourceType: 'javascript',
			lastModified: Date.now()
		};
	}

	/**
	 * Extract string value from an AST node
	 */
	private extractStringValue(node: any): string | null {
		if (node.type === 'StringLiteral') {
			return node.value;
		}
		return null;
	}

	/**
	 * Get property key name
	 */
	private getPropertyKey(prop: any): string | null {
		if (prop.key.type === 'Identifier') {
			return prop.key.name;
		}
		if (prop.key.type === 'StringLiteral') {
			return prop.key.value;
		}
		return null;
	}

	/**
	 * Extract state properties from an object expression
	 */
	private extractStateProperties(
		node: any,
		state: Map<string, PropertyInfo>
	): void {
		for (const prop of node.properties) {
			// Handle ObjectProperty (regular properties)
			if (prop.type === 'ObjectProperty') {
				const key = this.getPropertyKey(prop);
				if (!key) {
					continue;
				}

				const propertyInfo: PropertyInfo = {
					name: key,
					type: this.inferType(prop.value),
					sourceLine: prop.loc?.start?.line
				};

				// Check if value is an object
				if (prop.value && prop.value.type === 'ObjectExpression') {
					propertyInfo.isObject = true;
					propertyInfo.properties = new Map();
					this.extractStateProperties(prop.value, propertyInfo.properties);
				}

				state.set(key, propertyInfo);
			}
			// Handle ObjectMethod (getters only - regular methods don't belong in state)
			else if (prop.type === 'ObjectMethod' && prop.kind === 'get') {
				const key = this.getPropertyKey(prop);
				if (!key) {
					continue;
				}

				const propertyInfo: PropertyInfo = {
					name: key,
					type: 'computed', // Mark as computed property
					sourceLine: prop.loc?.start?.line
				};

				state.set(key, propertyInfo);
			}
			// Skip other ObjectMethod types (regular methods, setters)
		}
	}

	/**
	 * Extract methods from an object expression
	 */
	private extractMethods(node: any, methods: Map<string, MethodInfo>): void {
		for (const prop of node.properties) {
			const key = this.getPropertyKey(prop);
			if (!key) {
				continue;
			}

			// Skip getters and setters in actions/callbacks
			if (prop.type === 'ObjectMethod' && (prop.kind === 'get' || prop.kind === 'set')) {
				continue;
			}

			const methodInfo: MethodInfo = {
				name: key,
				sourceLine: prop.loc?.start?.line
			};

			// Extract parameters if this is a function
			if (
				prop.value &&
				(prop.value.type === 'FunctionExpression' ||
					prop.value.type === 'ArrowFunctionExpression')
			) {
				methodInfo.parameters = prop.value.params.map((param: any) => {
					if (param.type === 'Identifier') {
						return param.name;
					}
					return 'param';
				});
			} else if (prop.type === 'ObjectMethod') {
				methodInfo.parameters = prop.params.map((param: any) => {
					if (param.type === 'Identifier') {
						return param.name;
					}
					return 'param';
				});
			}

			methods.set(key, methodInfo);
		}
	}

	/**
	 * Infer type from a JavaScript value node
	 */
	private inferType(node: any): string | undefined {
		if (!node) {
			return undefined;
		}

		switch (node.type) {
			case 'StringLiteral':
				return 'string';
			case 'NumericLiteral':
				return 'number';
			case 'BooleanLiteral':
				return 'boolean';
			case 'ArrayExpression':
				return 'array';
			case 'ObjectExpression':
				return 'object';
			case 'NullLiteral':
				return 'null';
			case 'ArrowFunctionExpression':
			case 'FunctionExpression':
				return 'function';
			default:
				return undefined;
		}
	}
}
